File: backend/app/domain/slide_geometry.py
"""统一解析页面块几何：fixed 走布局槽位，flex 走布局树 solver。"""

from __future__ import annotations

from pydantic import BaseModel

from app.domain.content import Slide
from app.domain.geometry import Rect
from app.domain.layout import get_layout


class PlacedBlock(BaseModel):
    block_id: str
    rect: Rect
    # TextStyleName；用 str 避免与 layout 强耦合
    text_style: str | None = None


def resolve_slide_geometry(slide: Slide) -> list[PlacedBlock]:
    if slide.layout_mode == "flex":
        if slide.layout_tree is None:
            return []
        from app.domain.flex_solve import solve

        return solve(slide.layout_tree)

    layout = get_layout(slide.layout_id)
    result: list[PlacedBlock] = []
    for block in slide.blocks:
        slot = layout.slot_by_id(block.slot_id)
        if slot is not None:
            result.append(
                PlacedBlock(
                    block_id=block.id,
                    rect=slot.rect,
                    text_style=slot.text_style,
                )
            )
    return result


def placed_by_block_id(slide: Slide) -> dict[str, PlacedBlock]:
    return {p.block_id: p for p in resolve_slide_geometry(slide)}

File: java_backend/src/main/java/com/aippt/domain/flex/SlideGeometry.java
package com.aippt.domain.flex;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.aippt.domain.Geometry;
import com.aippt.domain.content.SlideContent;

public final class SlideGeometry {

    private SlideGeometry() {
    }

    public static List<FlexSolve.PlacedBlock> resolve(SlideContent slide, com.aippt.domain.Layout layout) {
        if ("flex".equals(slide.layoutMode())) {
            if (slide.layoutTree() == null) {
                return List.of();
            }
            return FlexSolve.solve(slide.layoutTree());
        }
        if (layout == null || layout.slots() == null) {
            return List.of();
        }
        List<FlexSolve.PlacedBlock> result = new ArrayList<>();
        for (Map<String, Object> block : slide.blocks()) {
            com.aippt.domain.Layout.Slot slot = layout.slotById(com.aippt.domain.content.Blocks.slotId(block));
            if (slot != null && slot.rect() != null) {
                result.add(new FlexSolve.PlacedBlock(
                        com.aippt.domain.content.Blocks.id(block),
                        slot.rect(),
                        slot.textStyle()
                ));
            }
        }
        return result;
    }

    public static Map<String, FlexSolve.PlacedBlock> placedByBlockId(
            SlideContent slide,
            com.aippt.domain.Layout layout
    ) {
        Map<String, FlexSolve.PlacedBlock> map = new HashMap<>();
        for (FlexSolve.PlacedBlock placed : resolve(slide, layout)) {
            map.put(placed.blockId(), placed);
        }
        return map;
    }
}
